### ISSUE TITLE
Numerical Instability in Softmax Activation Function Due to Overflow in PyTorch MLP Implementation

### ISSUE THREAD

**Issue by ShailaShova** [ethangilmore/micronet#1]
While using the MLP architecture and parameters provided in the example notebook (example.ipynb), I encountered an issue when applying the model to the make_blobs dataset from sklearn. Specifically, the loss function produces NaN values during training due to numerical instability in the softmax activation function. The overflow in the exp function within the softmax computation leads to this instability.

Dataset Used: make_blobs from sklearn.datasets
Architecture:
Input layer: 2 neurons
Hidden layers: 2 layers with 8 neurons each, using ReLU activation
Output layer: 2 neurons, using Softmax activation
Training Parameters:
Epochs: 20
Batch Size: 16
Learning Rate: 0.2

**Issue:**
The current implementation of the softmax function does not handle large input values well, leading to an overflow in the exponential calculation. As a result, the output of the softmax becomes NaN, which propagates through the loss function, making the training process unstable and resulting in NaN values for the loss.

**Suggested Fix**
To address this issue, I recommend using a numerically stable implementation of the softmax function, which involves shifting the input by the maximum value before exponentiation. 

def stable_softmax(x):
    shift_x = x - np.max(x)
    y = np.exp(shift_x) / np.sum(np.exp(shift_x))
    def backward(dy):
        J = -np.outer(y, y)
        np.fill_diagonal(J, y * (1 - y))
        return dy @ J
    return y, backward
